data_structures/src/radon_report.rs

```diff
@@ -0,0 +1,183 @@
+///
+/// It contains a RAD result paired with metadata specific to the stage of the script being executed.
+///
+/// `RT` is the generalization of `RadonTypes`, and `IE` is the generalization of `RadError`
+#[derive(Clone, Debug)]
+pub struct RadonReport<RT>
+where
+    RT: TypeLike,
+{
+    /// Stage-specific metadata.
+    pub metadata: Stage,
+    /// This is raw result: either a RadonTypes or a RadonError.
+    pub result: Result<RT, RadonError<RT::Error>>,
+    /// Keep track of how many milliseconds did the execution take to complete or fail.
+    pub running_time: Duration,
+}
+
+/// Implementations, factories and convenience methods for `RadonReport`.
+impl<RT> RadonReport<RT>
+where
+    RT: TypeLike,
+{
+    /// Factory for constructing a `RadonReport` from the `Result` of something that could be
+    /// `ErrorLike` plus a `ReportContext`.
+    pub fn from_result(
+        result: Result<RT, RT::Error>,
+        context: &ReportContext,
+    ) -> Result<Self, RT::Error> {
+        let result = RT::Error::intercept(result);
+
+        Ok(RadonReport {
+            result,
+            metadata: context.stage.clone(),
+            running_time: context.duration(),
+        })
+    }
+
+    /// Recover a `Result` in the likes of `Result<RadonTypes, RadError>` from a `RadonReport`.
+    pub fn into_inner(self) -> Result<RT, RT::Error> {
+        self.result
+            .map_err(|radon_error| radon_error.inner.unwrap_or_default())
+    }
+}
+
+/// This is the main serializer for turning `RadonReport` into a CBOR-encoded byte stream that can be
+/// consumed by any Witnet library, e.g. the `UsingWitnet` solidity contract.
+impl<RT> TryFrom<&RadonReport<RT>> for Vec<u8>
+where
+    RT: TypeLike,
+{
+    type Error = RT::Error;
+
+    fn try_from(report: &RadonReport<RT>) -> Result<Self, Self::Error> {
+        match report.result {
+            Ok(ref radon_types) => radon_types.encode(),
+            Err(ref radon_error) => radon_error.encode(),
+        }
+    }
+}
+
+/// This trait identifies a RADON-compatible type system, i.e. most likely an `enum` with different
+/// cases for different data types.
+pub trait TypeLike {
+    type Error: ErrorLike;
+
+    fn encode(&self) -> Result<Vec<u8>, Self::Error>;
+}
+
+/// A generic structure for bubbling up any kind of metadata that may be generated during the
+/// execution of a RADON script.
+#[derive(Default)]
+pub struct ReportContext {
+    pub call_arguments: Option<Vec<serde_cbor::Value>>,
+    pub call_argument_index: Option<u8>,
+    pub call_index: Option<u8>,
+    pub call_operator: Option<u8>,
+    pub stage: Stage,
+    pub completion_time: Option<Instant>,
+    pub start_time: Option<Instant>,
+    pub script_index: Option<u8>,
+}
+
+/// Implementation of convenience methods for `ReportContext`
+impl ReportContext {
+    /// Set start time.
+    pub fn start(&mut self) {
+        self.start_time = Some(Instant::now());
```

Comment:
should we use NTP time?

Reply:
No, this is only used to calculate a duration. We don't care about the exact start time.